File: legacy/QueKinhDich/Business/QueDich.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace KinhDich
{    
    /// <summary>
    /// Lớp mô tả một quẻ dịch
    /// </summary>
    public class QueDich
    {
        public Hao[] hao;
        public Dictionary<string, int> diemLucThan;

        public int soQueThuong, soQueHa;
        public int quebien;
        public string Cung;
        public string TenQueThuong, TenQueHa, TenQueDich;
        int HaoThe, HaoUng;
        NgayAmLich amlich;

        public QueDich QueDichBien;
        public QueDich QueChu;

        /// <summary>
        /// Khởi tạo một quẻ dịch
        /// </summary>
        public QueDich()
        {
            hao = new Hao[7];
            for (int i = 0; i < 7; i++)
                hao[i] = new Hao();
            quebien = 0;
            soQueThuong = 0; soQueHa = 0;
            diemLucThan = new Dictionary<string, int>();
            for (int i = 0; i < 5; i++)
                diemLucThan.Add(Const.lucthan[i], int.MinValue);
        }        
        /// <summary>
        /// Tạo một quẻ dịch với quẻ thượng và quẻ hạ có tên được xác định trước
        /// </summary>
        /// <param name="quethuong"></param>
        /// <param name="queha"></param>
        public QueDich(DateTime time, string quethuong, string queha)
        {
            amlich = new NgayAmLich(time);
            hao = new Hao[7];
            for (int i = 0; i < 7; i++)
                hao[i] = new Hao();
            diemLucThan = new Dictionary<string, int>();
            for (int i = 0; i < 5; i++)
                diemLucThan.Add(Const.lucthan[i], int.MinValue);
            //business.XacDinhQueKinhDich(time, out soQueThuong, out soQueHa, out quebien, out TenQueThuong, out TenQueHa);
            TenQueThuong = quethuong;
            TenQueHa = queha;
            quebien = 0;
            soQueThuong = business.vitriTienThien(quethuong);
            soQueHa = business.vitriTienThien(queha);

            TenQueDich = business.findTenQue6Hao(TenQueThuong, TenQueHa);
            Cung = business.findCungQueDich(TenQueThuong, TenQueHa);
            NapGiap();
        }

        /// <summary>
        /// Tạo quẻ dịch với thời gian cho trước (dương lịch)
        /// </summary>
        /// <param name="time"></param>
        public QueDich(DateTime time)
        {
            amlich = new NgayAmLich(time);
            hao = new Hao[7];
            for (int i = 0; i < 7; i++)
                hao[i] = new Hao();
            diemLucThan = new Dictionary<string, int>();
            for (int i = 0; i < 5; i++)
                diemLucThan.Add(Const.lucthan[i], int.MinValue);

            business.XacDinhQueKinhDich(time, out soQueThuong, out soQueHa, out quebien, out TenQueThuong, out TenQueHa);
            TenQueDich = business.findTenQue6Hao(TenQueThuong, TenQueHa);
            Cung = business.findCungQueDich(TenQueThuong, TenQueHa);
            NapGiap();
        }

        /// <summary>
        /// Tạo quẻ dịch cuộc đời
        /// </summary>
        /// <param name="time"></param>
        public QueDich(DateTime time,bool cuocdoi)
        {
            if (cuocdoi == false) return;

            amlich = new NgayAmLich(time);
            hao = new Hao[7];
            for (int i = 0; i < 7; i++)
                hao[i] = new Hao();
            diemLucThan = new Dictionary<string, int>();
            for (int i = 0; i < 5; i++)
                diemLucThan.Add(Const.lucthan[i], int.MinValue);

            business.XacDinhQueCuocDoi(time, out soQueThuong, out soQueHa, out quebien, out TenQueThuong, out TenQueHa);
            TenQueDich = business.findTenQue6Hao(TenQueThuong, TenQueHa);
            Cung = business.findCungQueDich(TenQueThuong, TenQueHa);
            NapGiap();
        }

        public void BienQue()
        {
            business.BienQue(soQueThuong, soQueHa, quebien, out TenQueThuong, out TenQueHa);
            TenQueDich = business.findTenQue6Hao(TenQueThuong, TenQueHa);

            soQueThuong = business.findSoQueTienThien(TenQueThuong);
            soQueHa = business.findSoQueTienThien(TenQueHa);
            quebien = 0;

            NapGiap();
        }

        private void TuanKhong()
        {
            string tuankhong1, tuankhong2;
            //Tinh Tuan Khong
            business.TuanKhong(amlich.ThienCanNgay, amlich.DiaChiNgay, out tuankhong1, out tuankhong2);

            for (int i = 1; i < 7; i++)
                hao[i].TuanKhong = false;
            for (int i = 1; i < 7; i++)
                if ((hao[i].chi == tuankhong1) || (hao[i].chi == tuankhong2))
                    hao[i].TuanKhong = true;

            for (int i = 1; i < 7; i++)
                if (hao[i].TuanKhong) hao[i].napgiap = hao[i].napgiap + "- Không";
        }

        private void Than()
        {

            string thanhao1 = business.findThan(amlich.ThienCanNgay);

            int j = 0;
            while (Const.than[j] != thanhao1) j++;

            for (int i = 1; i < 7; i++)
            {
                hao[i].than = Const.than[j];
                j++;
            }
        }

        /// <summary>
        /// Nap Giap cho cac Hao
        /// </summary>
        public void NapGiap()
        {
            business.findNapGiapQueThuong(Cung, TenQueThuong, ref hao[4], ref hao[5], ref hao[6]);
            business.findNapGiapQueHa(Cung, TenQueHa, ref hao[1], ref hao[2], ref hao[3]);

            //Tìm hào thế, ứng
            HaoThe = business.findHaoThe(TenQueDich);
            if (HaoThe == 6) HaoUng = 3;
            if (HaoThe == 5) HaoUng = 2;
            if (HaoThe == 4) HaoUng = 1;
            if (HaoThe == 3) HaoUng = 6;
            if (HaoThe == 2) HaoUng = 5;
            if (HaoThe == 1) HaoUng = 4;
            hao[HaoThe].napgiap = hao[HaoThe].napgiap + " (Thế)";
            hao[HaoThe].HaoThe = true;

            hao[HaoUng].napgiap = hao[HaoUng].napgiap + " (Ứng)";
            hao[HaoUng].HaoUng = true;

            //Tính Tuần Không
            TuanKhong();

            //Tính Lục Thần
            Than();
        }

        private void TinhDiemHao(string nguhanh, ref Hao hao)
        {
            if (nguhanh == hao.nguhanh)
                hao.diemso = hao.diemso + 2;
            else
                if (business.TuongSinh(nguhanh, hao.nguhanh))
                    hao.diemso = hao.diemso + 2;
                else
                    //Mat suc do sinh
                    if (business.TuongSinh(hao.nguhanh, nguhanh))
                        hao.diemso = hao.diemso - 1;
                    else
                        //Mat suc do khac
                        if (business.TuongKhac(hao.nguhanh, nguhanh))
                            hao.diemso = hao.diemso - 1;
                        //Bi ngay khac
                        else
                            if (business.TuongKhac(nguhanh, hao.nguhanh))
                                hao.diemso = hao.diemso - 2;
        }

        /// <summary>
        /// Xác định điểm số cho lục thân
        /// </summary>
        private void TinhDiemLucThan()
        {

            for (int i = 1; i < 7; i++)
            {
                int diemhienthoi = diemLucThan[hao[i].lucthan];
                diemLucThan[hao[i].lucthan] = Math.Max(diemhienthoi, hao[i].diemso);
            }

            Dictionary<string, int> temp = new Dictionary<string, int>(diemLucThan);

            foreach (var d in temp)
                if ((int)d.Value == int.MinValue)
                {
                    //Neu khong co trong hao, phai tinh diem dua tren que chu
                    //QueDich _quechu = new QueDich(amlich.NgayDuonglich,Cung, Cung);

                    //Lay diem so cua que chu
                    //diemLucThan.Add(d.Key,QueChu.diemLucThan[d.Key]);
                    diemLucThan.Remove(d.Key);
                    diemLucThan.Add(d.Key, QueChu.diemLucThan[d.Key]);
                }
        }

        private void TinhDiemHao(int isQueChu)
        {
            for (int i = 1; i < 7; i++)
            {
                //Xac dinh diem so cho tu Hao

                //Nhat Kien sinh tro
                string nguhanhNgay = business.findNguHanh(amlich.DiaChiNgay);
                TinhDiemHao(nguhanhNgay, ref hao[i]);

                //Nguyet Kien sinh tro
                string nguhanhThang = business.findNguHanh(amlich.DiaChiThang);
                TinhDiemHao(nguhanhThang, ref hao[i]);

                //Hao Bien sinh tro
                string nguhanhHaoDong = hao[quebien].nguhanh;
                TinhDiemHao(nguhanhHaoDong, ref hao[i]);

                if (isQueChu != 0)
                {
                    string nguhanhHaoBien = QueDichBien.hao[i].nguhanh;
                    TinhDiemHao(nguhanhHaoBien, ref hao[i]);
                }
            }
        }
        /// <summary>
        /// Tính độ vượng suy của một quẻ
        /// </summary>
        public void GiaiQue()
        {
            QueDichBien = new QueDich(amlich.NgayDuonglich);
            //QueDichBien = new QueDich(que);
            QueDichBien.BienQue();
            
            QueChu = new QueDich(amlich.NgayDuonglich, Cung, Cung);

            QueChu.TinhDiemHao(0);
            QueChu.TinhDiemLucThan();
            
            TinhDiemHao(1);
            TinhDiemLucThan();            
        }

        /// <summary>
        /// Tính độ vượng suy của một quẻ
        /// </summary>
        public void GiaiQueCuocDoi()
        {
            QueDichBien = new QueDich(amlich.NgayDuonglich,true);//Tinh theo que cuoc doi            
            QueDichBien.BienQue();

            QueChu = new QueDich(amlich.NgayDuonglich, Cung, Cung);

            QueChu.TinhDiemHao(0);
            QueChu.TinhDiemLucThan();

            TinhDiemHao(1);
            TinhDiemLucThan();
        }
    }
}

File: src/core/queDich.ts
/**
 * Port từ QueKinhDich/Business/QueDich.cs. Giữ nguyên cách đánh số hào 1-6 (mảng `hao` có
 * 7 phần tử, index 0 không dùng) để dễ đối chiếu trực tiếp với code C# gốc khi review.
 */
import { LUC_THAN, THAN } from "./const";
import { taoHaoMoi, type Hao } from "./types";
import { tinhAmLich, type AmLich } from "./lunar";
import * as business from "./business";

export class QueDich {
  hao: Hao[];
  diemLucThan: Record<string, number> = {};

  soQueThuong = 0;
  soQueHa = 0;
  queBien = 0;
  cung = "";
  tenQueThuong = "";
  tenQueHa = "";
  tenQueDich = "";

  private haoThe = 0;
  private haoUng = 0;
  private amLich: AmLich;

  queDichBien?: QueDich;
  queChu?: QueDich;

  /** Port từ constructor `QueDich(DateTime time)` — an quẻ theo thời điểm dương lịch. */
  constructor(time: Date);
  /** Port từ constructor `QueDich(DateTime time, string quethuong, string queha)`. */
  constructor(time: Date, queThuong: string, queHa: string);
  /** Port từ constructor `QueDich(DateTime time, bool cuocdoi)` — quẻ Cuộc Đời. */
  constructor(time: Date, cuocDoi: true);
  constructor(time: Date, arg2?: string | true, arg3?: string) {
    this.hao = [taoHaoMoi(), taoHaoMoi(), taoHaoMoi(), taoHaoMoi(), taoHaoMoi(), taoHaoMoi(), taoHaoMoi()];
    for (const lt of LUC_THAN) this.diemLucThan[lt] = Number.NEGATIVE_INFINITY;
    this.amLich = tinhAmLich(time);

    if (typeof arg2 === "string" && typeof arg3 === "string") {
      // QueDich(time, quethuong, queha)
      this.tenQueThuong = arg2;
      this.tenQueHa = arg3;
      this.queBien = 0;
      this.soQueThuong = business.vitriTienThien(arg2);
      this.soQueHa = business.vitriTienThien(arg3);
      this.tenQueDich = business.findTenQue6Hao(this.tenQueThuong, this.tenQueHa);
      this.cung = business.findCungQueDich(this.tenQueThuong, this.tenQueHa);
      this.napGiap();
      return;
    }

    if (arg2 === true) {
      // QueDich(time, cuocdoi)
      const kq = business.xacDinhQueCuocDoi(time, this.amLich);
      this.soQueThuong = kq.soQueThuong;
      this.soQueHa = kq.soQueHa;
      this.queBien = kq.queBien;
      this.tenQueThuong = kq.tenQueThuong;
      this.tenQueHa = kq.tenQueHa;
      this.tenQueDich = business.findTenQue6Hao(this.tenQueThuong, this.tenQueHa);
      this.cung = business.findCungQueDich(this.tenQueThuong, this.tenQueHa);
      this.napGiap();
      return;
    }

    // QueDich(time)
    const kq = business.xacDinhQueKinhDich(time, this.amLich);
    this.soQueThuong = kq.soQueThuong;
    this.soQueHa = kq.soQueHa;
    this.queBien = kq.queBien;
    this.tenQueThuong = kq.tenQueThuong;
    this.tenQueHa = kq.tenQueHa;
    this.tenQueDich = business.findTenQue6Hao(this.tenQueThuong, this.tenQueHa);
    this.cung = business.findCungQueDich(this.tenQueThuong, this.tenQueHa);
    this.napGiap();
  }

  /** Port từ QueDich.BienQue() */
  bienQue(): void {
    const { queBienThuong, queBienHa } = business.bienQue(this.soQueThuong, this.soQueHa, this.queBien);
    this.tenQueThuong = queBienThuong;
    this.tenQueHa = queBienHa;
    this.tenQueDich = business.findTenQue6Hao(this.tenQueThuong, this.tenQueHa);

    this.soQueThuong = business.findSoQueTienThien(this.tenQueThuong) - 1;
    this.soQueHa = business.findSoQueTienThien(this.tenQueHa) - 1;
    this.queBien = 0;

    this.napGiap();
  }

  private tinhTuanKhong(): void {
    const [tk1, tk2] = business.tuanKhong(this.amLich.thienCanNgay, this.amLich.diaChiNgay);
    for (let i = 1; i < 7; i++) this.hao[i].tuanKhong = false;
    for (let i = 1; i < 7; i++) {
      if (this.hao[i].chi === tk1 || this.hao[i].chi === tk2) this.hao[i].tuanKhong = true;
    }
    for (let i = 1; i < 7; i++) {
      if (this.hao[i].tuanKhong) this.hao[i].napgiap += "- Không";
    }
  }

  private tinhThan(): void {
    const thanHao1 = business.findThan(this.amLich.thienCanNgay);
    let j = THAN.indexOf(thanHao1 as (typeof THAN)[number]);
    for (let i = 1; i < 7; i++) {
      this.hao[i].than = THAN[j];
      j++;
    }
  }

  /** Port từ QueDich.NapGiap() */
  napGiap(): void {
    business.napGiapQueThuong(this.cung, this.tenQueThuong, this.hao[4], this.hao[5], this.hao[6]);
    business.napGiapQueHa(this.cung, this.tenQueHa, this.hao[1], this.hao[2], this.hao[3]);

    this.haoThe = business.findHaoThe(this.tenQueDich);
    const map: Record<number, number> = { 6: 3, 5: 2, 4: 1, 3: 6, 2: 5, 1: 4 };
    this.haoUng = map[this.haoThe];

    this.hao[this.haoThe].napgiap += " (Thế)";
    this.hao[this.haoThe].haoThe = true;
    this.hao[this.haoUng].napgiap += " (Ứng)";
    this.hao[this.haoUng].haoUng = true;

    this.tinhTuanKhong();
    this.tinhThan();
  }

  private tinhDiemMotHao(nguHanh: string, hao: Hao): void {
    if (nguHanh === hao.nguhanh) {
      hao.diemso += 2;
    } else if (business.TuongSinh(nguHanh, hao.nguhanh)) {
      hao.diemso += 2;
    } else if (business.TuongSinh(hao.nguhanh, nguHanh)) {
      hao.diemso -= 1; // mất sức do sinh
    } else if (business.TuongKhac(hao.nguhanh, nguHanh)) {
      hao.diemso -= 1; // mất sức do khắc
    } else if (business.TuongKhac(nguHanh, hao.nguhanh)) {
      hao.diemso -= 2; // bị ngày khắc
    }
  }

  private tinhDiemLucThan(): void {
    for (let i = 1; i < 7; i++) {
      const hienThoi = this.diemLucThan[this.hao[i].lucthan];
      this.diemLucThan[this.hao[i].lucthan] = Math.max(hienThoi, this.hao[i].diemso);
    }

    for (const [key, value] of Object.entries({ ...this.diemLucThan })) {
      if (value === Number.NEGATIVE_INFINITY) {
        // Không có hào nào mang Lục Thân này — lấy điểm từ Quẻ Chủ
        this.diemLucThan[key] = this.queChu!.diemLucThan[key];
      }
    }
  }

  private tinhDiemHao(laQueChu: boolean): void {
    for (let i = 1; i < 7; i++) {
      const nguHanhNgay = business.findNguHanh(this.amLich.diaChiNgay);
      this.tinhDiemMotHao(nguHanhNgay, this.hao[i]);

      const nguHanhThang = business.findNguHanh(this.amLich.diaChiThang);
      this.tinhDiemMotHao(nguHanhThang, this.hao[i]);

      // hao[0] là phần tử placeholder (nguhanh="") — khi queBien=0 (trường hợp Quẻ Chủ),
      // hành vi này khớp với bản gốc: hao[quebien].nguhanh trả về chuỗi rỗng.
      const nguHanhHaoDong = this.hao[this.queBien].nguhanh;
      this.tinhDiemMotHao(nguHanhHaoDong, this.hao[i]);

      if (laQueChu) {
        const nguHanhHaoBien = this.queDichBien!.hao[i].nguhanh;
        this.tinhDiemMotHao(nguHanhHaoBien, this.hao[i]);
      }
    }
  }

  /** Port từ QueDich.GiaiQue() — tính điểm vượng suy cho từng Lục Thân của quẻ. */
  giaiQue(): void {
    this.queDichBien = new QueDich(this.amLich.ngayDuongLich);
    this.queDichBien.bienQue();

    this.queChu = new QueDich(this.amLich.ngayDuongLich, this.cung, this.cung);
    this.queChu.tinhDiemHao(false);
    this.queChu.tinhDiemLucThan();

    this.tinhDiemHao(true);
    this.tinhDiemLucThan();
  }

  /** Port từ QueDich.GiaiQueCuocDoi() */
  giaiQueCuocDoi(): void {
    this.queDichBien = new QueDich(this.amLich.ngayDuongLich, true);
    this.queDichBien.bienQue();

    this.queChu = new QueDich(this.amLich.ngayDuongLich, this.cung, this.cung);
    this.queChu.tinhDiemHao(false);
    this.queChu.tinhDiemLucThan();

    this.tinhDiemHao(true);
    this.tinhDiemLucThan();
  }
}
